ansible/tests/enterprise/consul_namespaces_test.yml

```
@@ -0,0 +1,159 @@
+  register: create_namespaces
+  environment:
+    CONSUL_HTTP_TOKEN: master
+  failed_when: create_namespaces.rc > 0
+
+
+# 1.3 Generate tokens
+- name: Create administrator token for team1
+  shell: >
+    consul acl token create \
+    -format=json \
+    -namespace=team1 \
+    -description="Team1 administrator" \
+    -policy-name="namespace-management" \
+    | jq -r .SecretID
+  register: team1_namespace_admin_token
+  environment:
+    CONSUL_HTTP_TOKEN: master
+
+- name: Create administrator token for team2
+  shell: >
+    consul acl token create \
+    -format=json \
+    -namespace=team2 \
+    -description="Team2 administrator" \
+    -policy-name="namespace-management" \
+    | jq -r .SecretID
+  register: team2_namespace_admin_token
+  environment:
+    CONSUL_HTTP_TOKEN: master
+
+# 2.1. Verify read/write access, try list/register services
+- name: Try to read namespace-management policy namespace team1 with admin token team1 - success
+  shell: consul acl policy read -namespace=team1 -name=namespace-management
+  register: read_policy_try_1
+  environment:
+    CONSUL_HTTP_TOKEN: "{{ team1_namespace_admin_token.stdout }}"
+  failed_when: read_policy_try_1.rc > 0
+
+- name: Try to read namespace-management policy namespace team2 with admin token team1 - should fail
+  shell: consul acl policy read -namespace=team2 -name=namespace-management
+  register: read_policy_try_2
+  environment:
+    CONSUL_HTTP_TOKEN: "{{ team1_namespace_admin_token.stdout }}"
+  failed_when: read_policy_try_2.rc == 0
+
+- name: Try to read namespace-management policy namespace team2 without any token - should fail
+  shell: consul acl policy read -namespace=team2 -name=namespace-management
+  register: read_policy_try_3
+  environment:
+    CONSUL_HTTP_TOKEN: ""
+  failed_when: read_policy_try_3.rc == 0
+
+- name: Try register service 'some-web' in namespace=default with admin token team1 - should fail
+  shell: consul services register -namespace=default -name=some-web -address=1.1.1.1 -port=80
+  register: register_service_try_2
+  environment:
+    CONSUL_HTTP_TOKEN: "{{ team1_namespace_admin_token.stdout }}"
+  failed_when: register_service_try_2.rc == 0
+
+- name: Debug register_service_try_2
+  debug:
+    msg: "register_service_try_2 {{ register_service_try_2 }}"
+
+- name: Try register service 'some-web' in namespace=team2 with admin token team1 - should fail
+  shell: consul services register -namespace=team2 -name=some-web -address=1.1.1.1 -port=80
+  register: register_service_try_3
+  environment:
+    CONSUL_HTTP_TOKEN: "{{ team1_namespace_admin_token.stdout }}"
+  failed_when: register_service_try_3.rc == 0
+
+- name: Debug register_service_try_3
+  debug:
+    msg: "register_service_try_3 {{ register_service_try_3 }}"
+
+- name: Try register service 'some-web' in namespace=team1 with admin token team1 - success
+  shell: consul services register -namespace=team1 -name=some-web -address=1.1.1.1 -port=80
```

```suggestion
  shell: consul services register  -name=some-web -address=1.1.1.1 -port=80
```
I guess this would also work fine but may be we want to explicitly specify the namespace anyway?

---

I had to specify, service name and namespaces where to register it. If I do not specify `-namespace=team1`, it will be registered in `default` namespace.

---

Oh ok.
I read this on the hashicorp consul site so  wondered [namespace-inheritance-with-tokens](https://learn.hashicorp.com/tutorials/consul/namespaces-secure-shared-access#namespace-inheritance-with-tokens) :
```
A token's namespace can be inherited during requests related to services, intentions, Consul KV, checks, and ACLs. This means that if you register a service with a token in the app-team namespace, the service will be registered in that namespace without having to specify it explicitly.
```
Probably i understood it wrong!

---

No, you actually right. I missed that. Thank you. I would leave ansible script as it is (make it explicit)

---

:+1: